File: platforms/ios/Sources/WhiskerRuntime/paint/ClipPath.swift
import UIKit
import WhiskerModule

enum HostClipReferenceBox {
    case border
    case padding
    case content
}

enum HostClipPath {
    case inset(HostInsetClipPath)
    case circle(HostCircleClipPath)
    case ellipse(HostEllipseClipPath)
    case path(HostPathClipPath)

    var fillRule: CAShapeLayerFillRule {
        if case let .path(shape) = self, shape.evenOdd { return .evenOdd }
        return .nonZero
    }

    func path(in bounds: CGRect, contentBox: CGRect, painter: HostBoxPainter) -> CGPath {
        switch self {
        case let .inset(shape): shape.path(in: bounds, contentBox: contentBox, painter: painter)
        case let .circle(shape): shape.path(in: bounds, contentBox: contentBox, painter: painter)
        case let .ellipse(shape): shape.path(in: bounds, contentBox: contentBox, painter: painter)
        case let .path(shape): shape.path(in: bounds, contentBox: contentBox, painter: painter)
        }
    }
}

struct HostPathCommand {
    let kind: UInt32
    let points: [WhiskerMobileLengthPercentage]
}

struct HostPathClipPath {
    let referenceBox: HostClipReferenceBox
    let evenOdd: Bool
    let commands: [HostPathCommand]

    func path(in bounds: CGRect, contentBox: CGRect, painter: HostBoxPainter) -> CGPath {
        let reference = clipReferenceBox(referenceBox, bounds, contentBox, painter)
        func point(_ command: HostPathCommand, _ offset: Int) -> CGPoint {
            CGPoint(
                x: reference.minX + resolve(command.points[offset], axis: reference.width),
                y: reference.minY + resolve(command.points[offset + 1], axis: reference.height)
            )
        }
        let path = CGMutablePath()
        for command in commands {
            switch command.kind {
            case UInt32(WHISKER_PATH_MOVE_TO): path.move(to: point(command, 0))
            case UInt32(WHISKER_PATH_LINE_TO): path.addLine(to: point(command, 0))
            case UInt32(WHISKER_PATH_QUADRATIC_TO):
                path.addQuadCurve(to: point(command, 2), control: point(command, 0))
            case UInt32(WHISKER_PATH_CUBIC_TO):
                path.addCurve(
                    to: point(command, 4),
                    control1: point(command, 0),
                    control2: point(command, 2)
                )
            case UInt32(WHISKER_PATH_CLOSE): path.closeSubpath()
            default: break
            }
        }
        return path
    }
}

/// Rounded inset basic shape copied out of the borrowed mobile ABI payload.
struct HostInsetClipPath {
    let referenceBox: HostClipReferenceBox
    let edges: [WhiskerMobileLengthPercentage]
    let radiiHorizontal: [WhiskerMobileLengthPercentage]
    let radiiVertical: [WhiskerMobileLengthPercentage]

    func path(in bounds: CGRect, contentBox: CGRect, painter: HostBoxPainter) -> CGPath {
        let reference = switch referenceBox {
        case .border: bounds
        case .padding: painter.paddingBoxRect(in: bounds)
        case .content: contentBox
        }
        let top = resolve(edges[0], axis: reference.height)
        let right = resolve(edges[1], axis: reference.width)
        let bottom = resolve(edges[2], axis: reference.height)
        let left = resolve(edges[3], axis: reference.width)
        let inset = CGRect(
            x: reference.minX + left,
            y: reference.minY + top,
            width: max(0, reference.width - left - right),
            height: max(0, reference.height - top - bottom)
        )
        let radii = radiiHorizontal.indices.map { index in
            CGSize(
                width: max(0, resolve(radiiHorizontal[index], axis: inset.width)),
                height: max(0, resolve(radiiVertical[index], axis: inset.height))
            )
        }
        return roundedPath(in: inset, radii: radii).cgPath
    }
}

struct HostCircleClipPath {
    let referenceBox: HostClipReferenceBox
    let radius: WhiskerMobileLengthPercentage
    let centerX: WhiskerMobileLengthPercentage
    let centerY: WhiskerMobileLengthPercentage

    func path(in bounds: CGRect, contentBox: CGRect, painter: HostBoxPainter) -> CGPath {
        let reference = clipReferenceBox(referenceBox, bounds, contentBox, painter)
        let center = CGPoint(
            x: reference.minX + resolve(centerX, axis: reference.width),
            y: reference.minY + resolve(centerY, axis: reference.height)
        )
        let diagonal = hypot(reference.width, reference.height) / sqrt(2)
        let radius = max(0, resolve(radius, axis: diagonal))
        return UIBezierPath(arcCenter: center, radius: radius, startAngle: 0, endAngle: .pi * 2, clockwise: true).cgPath
    }
}

struct HostEllipseClipPath {
    let referenceBox: HostClipReferenceBox
    let radiusX: WhiskerMobileLengthPercentage
    let radiusY: WhiskerMobileLengthPercentage
    let centerX: WhiskerMobileLengthPercentage
    let centerY: WhiskerMobileLengthPercentage

    func path(in bounds: CGRect, contentBox: CGRect, painter: HostBoxPainter) -> CGPath {
        let reference = clipReferenceBox(referenceBox, bounds, contentBox, painter)
        let centerX = reference.minX + resolve(centerX, axis: reference.width)
        let centerY = reference.minY + resolve(centerY, axis: reference.height)
        let radiusX = max(0, resolve(radiusX, axis: reference.width))
        let radiusY = max(0, resolve(radiusY, axis: reference.height))
        return UIBezierPath(ovalIn: CGRect(
            x: centerX - radiusX, y: centerY - radiusY,
            width: radiusX * 2, height: radiusY * 2
        )).cgPath
    }
}

private func clipReferenceBox(
    _ box: HostClipReferenceBox,
    _ bounds: CGRect,
    _ contentBox: CGRect,
    _ painter: HostBoxPainter
) -> CGRect {
    switch box {
    case .border: bounds
    case .padding: painter.paddingBoxRect(in: bounds)
    case .content: contentBox
    }
}

private func resolve(_ value: WhiskerMobileLengthPercentage, axis: CGFloat) -> CGFloat {
    CGFloat(value.length) + CGFloat(value.fraction) * axis
}

File: platforms/android/runtime/src/main/kotlin/rs/whisker/runtime/paint/ClipPath.kt
package rs.whisker.runtime.paint

import android.graphics.Path
import android.graphics.RectF

internal enum class HostClipReferenceBox { Border, Padding, Content }

internal sealed interface HostClipPath { val referenceBox: HostClipReferenceBox }

/** Rounded inset basic shape retained in physical Host coordinates. */
internal data class HostInsetClipPath(
    override val referenceBox: HostClipReferenceBox,
    val edges: List<HostPaintCoordinate>,
    val radiiHorizontal: List<HostPaintCoordinate>,
    val radiiVertical: List<HostPaintCoordinate>,
) : HostClipPath

internal data class HostCircleClipPath(
    override val referenceBox: HostClipReferenceBox,
    val radius: HostPaintCoordinate,
    val centerX: HostPaintCoordinate,
    val centerY: HostPaintCoordinate,
) : HostClipPath

internal data class HostEllipseClipPath(
    override val referenceBox: HostClipReferenceBox,
    val radiusX: HostPaintCoordinate,
    val radiusY: HostPaintCoordinate,
    val centerX: HostPaintCoordinate,
    val centerY: HostPaintCoordinate,
) : HostClipPath

internal data class HostPathCommand(
    val kind: Int,
    val points: List<HostPaintCoordinate>,
)

internal data class HostPathClipPath(
    override val referenceBox: HostClipReferenceBox,
    val evenOdd: Boolean,
    val commands: List<HostPathCommand>,
) : HostClipPath

internal fun resolveClipPath(
    clip: HostClipPath,
    width: Float,
    height: Float,
    borderWidths: FloatArray,
    contentBox: RectF,
): Path {
    val reference = when (clip.referenceBox) {
        HostClipReferenceBox.Border -> RectF(0f, 0f, width, height)
        HostClipReferenceBox.Padding -> RectF(
            borderWidths[3].coerceIn(0f, width),
            borderWidths[0].coerceIn(0f, height),
            width - borderWidths[1].coerceIn(0f, width),
            height - borderWidths[2].coerceIn(0f, height),
        )
        HostClipReferenceBox.Content -> RectF(contentBox)
    }
    if (clip is HostPathClipPath) {
        fun point(command: HostPathCommand, offset: Int) = android.graphics.PointF(
            reference.left + command.points[offset].resolve(reference.width()),
            reference.top + command.points[offset + 1].resolve(reference.height()),
        )
        return Path().apply {
            fillType = if (clip.evenOdd) Path.FillType.EVEN_ODD else Path.FillType.WINDING
            clip.commands.forEach { command ->
                when (command.kind) {
                    PATH_MOVE_TO -> point(command, 0).also { moveTo(it.x, it.y) }
                    PATH_LINE_TO -> point(command, 0).also { lineTo(it.x, it.y) }
                    PATH_QUADRATIC_TO -> {
                        val control = point(command, 0)
                        val end = point(command, 2)
                        quadTo(control.x, control.y, end.x, end.y)
                    }
                    PATH_CUBIC_TO -> {
                        val control1 = point(command, 0)
                        val control2 = point(command, 2)
                        val end = point(command, 4)
                        cubicTo(control1.x, control1.y, control2.x, control2.y, end.x, end.y)
                    }
                    PATH_CLOSE -> close()
                }
            }
        }
    }
    if (clip is HostCircleClipPath) {
        val centerX = reference.left + clip.centerX.resolve(reference.width())
        val centerY = reference.top + clip.centerY.resolve(reference.height())
        val diagonal = kotlin.math.hypot(reference.width(), reference.height()) / kotlin.math.sqrt(2f)
        val radius = clip.radius.resolve(diagonal).coerceAtLeast(0f)
        return Path().apply { addCircle(centerX, centerY, radius, Path.Direction.CW) }
    }
    if (clip is HostEllipseClipPath) {
        val centerX = reference.left + clip.centerX.resolve(reference.width())
        val centerY = reference.top + clip.centerY.resolve(reference.height())
        val radiusX = clip.radiusX.resolve(reference.width()).coerceAtLeast(0f)
        val radiusY = clip.radiusY.resolve(reference.height()).coerceAtLeast(0f)
        return Path().apply {
            addOval(RectF(centerX - radiusX, centerY - radiusY, centerX + radiusX, centerY + radiusY), Path.Direction.CW)
        }
    }
    clip as HostInsetClipPath
    val top = clip.edges[0].resolve(reference.height())
    val right = clip.edges[1].resolve(reference.width())
    val bottom = clip.edges[2].resolve(reference.height())
    val left = clip.edges[3].resolve(reference.width())
    val inset = RectF(
        reference.left + left,
        reference.top + top,
        (reference.right - right).coerceAtLeast(reference.left + left),
        (reference.bottom - bottom).coerceAtLeast(reference.top + top),
    )
    if (inset.isEmpty) return Path()
    val radii = normalizeRadii(
        floatArrayOf(
            clip.radiiHorizontal[0].resolve(inset.width()).coerceAtLeast(0f),
            clip.radiiVertical[0].resolve(inset.height()).coerceAtLeast(0f),
            clip.radiiHorizontal[1].resolve(inset.width()).coerceAtLeast(0f),
            clip.radiiVertical[1].resolve(inset.height()).coerceAtLeast(0f),
            clip.radiiHorizontal[2].resolve(inset.width()).coerceAtLeast(0f),
            clip.radiiVertical[2].resolve(inset.height()).coerceAtLeast(0f),
            clip.radiiHorizontal[3].resolve(inset.width()).coerceAtLeast(0f),
            clip.radiiVertical[3].resolve(inset.height()).coerceAtLeast(0f),
        ),
        inset.width(),
        inset.height(),
    )
    return Path().apply { addRoundRect(inset, radii, Path.Direction.CW) }
}

internal const val PATH_MOVE_TO = 0
internal const val PATH_LINE_TO = 1
internal const val PATH_QUADRATIC_TO = 2
internal const val PATH_CUBIC_TO = 3
internal const val PATH_CLOSE = 4
